File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        int n = s.length();
        if (t.length() > n) {
            return "";
        }
        
        int i = 0, j = 0, minWindowSize = Integer.MAX_VALUE, reqCnt = t.length(), start = 0;
        HashMap<Character, Integer> freq = new HashMap<>();
        
        for (char ch : t.toCharArray()) {
            freq.put(ch, freq.getOrDefault(ch, 0) + 1);
        }
        
        while (j < n) {
            char ch = s.charAt(j);
            if (freq.getOrDefault(ch, 0) > 0) {
                reqCnt--;
            }
            freq.put(ch, freq.getOrDefault(ch, 0) - 1);
            
            while (reqCnt == 0) {
                int curWindowSize = j - i + 1;
                if (curWindowSize < minWindowSize) {
                    minWindowSize = curWindowSize;
                    start = i;
                }
                
                freq.put(s.charAt(i), freq.getOrDefault(s.charAt(i), 0) + 1);
                if (freq.get(s.charAt(i)) > 0) {
                    reqCnt++;
                }
                
                i++;
            }
            
            j++;
        }
        return minWindowSize == Integer.MAX_VALUE ? "" : s.substring(start, start + minWindowSize);
    }
}

File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        int S = s.size(), T = t.size();
        
        if (T > S) {
            return "";
        }
        
        unordered_map<int, int> mp;
        
        for (auto& it : t) {
            mp[it]++;
        }
        
        int start = -1, min_len = INT_MAX;
        
        for (int l = 0, r = 0, cnt = 0; r < S; r++) {
            if (mp[s[r]] > 0) {
                cnt++;
            }
            mp[s[r]]--;
            
            while (cnt == T) {
                int cur_window = r - l + 1;
                if (cur_window < min_len) {
                    start = l;
                    min_len = cur_window;
                }
                mp[s[l]]++;
                if (mp[s[l]] > 0) {
                    cnt--;
                }
                l++;
            }
        }
        
        return start == -1 ? "" : s.substr(start, min_len);
    }
};